File: bots/telegram/telegram-bot.ts
import config = require("../../bot-manager/utils/config");

import { Telegraf } from "telegraf";
import * as middleware from "./bot-middleware";
import * as inlineKeyboard from "./inline-keyboard";
import { adminCommandHandlers } from "./admin-commands";
import { getChatIdsThatSubscribedOneOfChannelIds } from "./chat-data-persistence";
import channels = require("../../bot-manager/utils/channels");
import { i18n } from "./i18n";


class TelegramBot {
  private bot: any;
  private HELP_TEXT_ADMIN = "Following admin commands/messages are available:\n";

  constructor() {
    this.bot = new Telegraf(config.bots.telegram.TELEGRAM_BOT_TOKEN);

    this.bot.use(middleware.setIsAdmin);
    this.bot.use(middleware.chatDataPersistence);
    this.registerAdminCommands();
    this.registerInlineKeyboardCallbackHandler();

    this.bot.command("/channels", async (ctx) => {
      ctx.reply("Channels:", inlineKeyboard.channelsForChatAsReplyMarkup(ctx.chat_data.channelIds || []));
    });
    this.bot.command("/mychannels", async (ctx) => {
      if (!ctx.chat_data.channelIds || ctx.chat_data.channelIds.length === 0) {
        ctx.reply(i18n.you_dont_have_any_subscribed_channels_yet(ctx.chat_data.language));
      } else {
        ctx.reply(
          i18n.your_channels(ctx.chat_data.language),
          inlineKeyboard.channelsForChatAsReplyMarkup(ctx.chat_data.channelIds, {
            onlySubscribedChannels: true,
          })
        );
      }
    });

    this.bot.command("/language", async (ctx) => {
      ctx.reply(
        i18n.languages(ctx.chat_data.language),
        inlineKeyboard.languagesForChatAsReplyMarkup(ctx.chat_data.language)
      );
    });

    this.bot.telegram.setMyCommands([
      {
        command: "channels",
        description: "List all available channels",
      },
      {
        command: "mychannels",
        description: "List all your channels",
      },
      {
        command: "help",
        description: "Help",
      },
    ]);

    this.bot.start(this.sendHelp.bind(this));
    this.bot.help(this.sendHelp.bind(this));

    this.bot.launch();

    this.bot.telegram.getMe().then((botInfo) => {
      console.log(`Telegram Bot @${botInfo.username} started`);
    });
  }

  sendMessageToChannels(message: string, channelIds: Array<Number | String>): number {
    const chatIds = getChatIdsThatSubscribedOneOfChannelIds(channelIds);
    for (const chatId of chatIds) {
      this.bot.telegram.sendMessage(chatId, `${inlineKeyboard.channelIdsToChannelList(channelIds)}\n${message}`);
    }
    return chatIds.length;
  }

  private registerAdminCommands() {
    for (const command in adminCommandHandlers) {
      const argsHelpText = adminCommandHandlers[command].args.map((arg) => " <" + arg + ">").join("");
      this.HELP_TEXT_ADMIN += `\n*${command}${argsHelpText}*:\n${adminCommandHandlers[command].helpText}\n`;
    }
    this.bot.on("text", async (ctx, next) => {
      if (ctx.is_admin === true) {
        const words = ctx.message.text.split(" ");
        const command = words[0].toLowerCase();
        if (command in adminCommandHandlers) {
          await adminCommandHandlers[command].handler(ctx, command);
        }
      }
      next();
    });
  }

  private registerInlineKeyboardCallbackHandler() {
    for (const inlineKeyboardCommandHandler of inlineKeyboard.inlineKeyboardCommandHandlers) {
      this.bot.action(inlineKeyboardCommandHandler.regex, async (ctx, next) => {
        const callbackData = ctx.update.callback_query.data;
        const regexMatch = inlineKeyboardCommandHandler.regex.exec(callbackData);
        if (regexMatch === null) {
          this.notifyDevelopersAboutError(
            `Failed to get matches for callback data ${callbackData} using regex ${inlineKeyboardCommandHandler.regex}`
          );
          return next();
        }
        await inlineKeyboardCommandHandler.handler(ctx, callbackData, regexMatch.slice(1));
        next();
      });
    }
  }

  private async sendHelp(ctx) {
    await ctx.replyWithMarkdown(i18n.HELP_TEXT(ctx.chat_data.language));
    if (ctx.is_admin && this.HELP_TEXT_ADMIN) {
      ctx.replyWithMarkdown(this.HELP_TEXT_ADMIN);
    }
  }

  private notifyDevelopersAboutError(message: string): void {
    for (const chatId of config.bots.telegram.TELEGRAM_CHAT_IDS_NOTIFY_ON_ERROR) {
      this.bot.telegram.sendMessage(chatId, message);
    }
  }
}

export const bot = new TelegramBot();

File: bots/telegram/telegram_bot.py
from typing import List
import os
import sys
import traceback
from threading import Thread

from telegram import ParseMode
from telegram.ext import Updater, Defaults, CommandHandler, MessageHandler, Filters

from bots.telegram.chat_persistence import ChatPersistence
from bots.telegram.admin_handlers import getAdminUsernames, getAdminMessageHandlers, getDevUserIdsToNotifyOnError
from bots.messenger_bot import MessengerBot


class TelegramBot(MessengerBot):
	HELP_TEXT = """Hello, this Telegram Bot will send notifications to different channels. You can subscribe the channels relevant for you by listing all channels with /channels and then select the ones you want. Click again to unsubscribe.
To view only a list of channels you subscribed to, use /mychannels.

You may want to subscribe the General channel. Use /channels to get started."""
	ADMIN_HELP_TEXT = "Following admin commands/messages are available:\n"

	def __init__(self):
		super().__init__()

		self.updater = Updater(
			token=os.getenv("TELEGRAM_BOT_TOKEN"),
			persistence=ChatPersistence("./bots/telegram/telegram-chats.json"),
			defaults=Defaults(parse_mode=ParseMode.MARKDOWN),
			use_context=True
		)
		self.dispatcher = self.updater.dispatcher
		self.adminUsernames = getAdminUsernames()
		self.commands = [
			{
				"command": "channels",
				"description": "List all available channels",
				"handler": self.notYetImplemented,
			},
			{
				"command": "mychannels",
				"description": "List all your channels",
				"handler": self.notYetImplemented,
			},
			{
				"command": "help",
				"description": "Help",
				"handler": self.help,
			},
		]

		self.addCommandHandlers()
		self.addAdminMessageHandlers()
		self.addErrorHandler()

		self.updater.start_polling()
		print("Telegram Bot @"+ self.updater.bot.username + " started")

	def stop(self):
		self.updater.stop()

	def addCommandHandlers(self):
		self.dispatcher.add_handler(CommandHandler('start', self.help))
		for cmd in self.commands:
			self.dispatcher.add_handler(CommandHandler(cmd["command"], cmd["handler"]))

		self.updater.bot.set_my_commands([(cmd["command"], cmd["description"]) for cmd in self.commands])

	def help(self, update, context):
		update.message.reply_text(self.HELP_TEXT)

		if (update.message.from_user.username in self.adminUsernames):
			update.message.reply_text(self.ADMIN_HELP_TEXT)

	def notYetImplemented(self, update, context):
		update.message.reply_text("This command is not yet implemented.")


	def addAdminMessageHandlers(self):
		for handler in getAdminMessageHandlers():
			self.dispatcher.add_handler(MessageHandler(Filters.user(username=self.adminUsernames) & handler.filter(), handler.handler))
			self.ADMIN_HELP_TEXT += "\n" + handler.getHelpText() + "\n"

	def addErrorHandler(self):
		self.dispatcher.add_error_handler(self.errorHandler)

	def errorHandler(self, update, context):
		# https://github.com/python-telegram-bot/python-telegram-bot/wiki/Code-snippets#an-good-error-handler
		if update.effective_message:
			update.effective_message.reply_text("An error occurred. The developers have been notified and are working on it.")

		message = "An Error occurred: " + str(context.error)
		trace = traceback.format_tb(sys.exc_info()[2])
		print(str(context.error) + "\n" + "".join(trace))
		for dev in getDevUserIdsToNotifyOnError():
			context.bot.send_message(dev, message + "\n" + str(trace[-1]), parse_mode=None)

	def sendMessageToChannels(self, message: str, channels: List[str]) -> int:
		print("send message", message, channels)
		return 0
